istio/pilot
# add webhook validation

**What this PR does / why we need it**:

Add the `istio.io/pilot/platform/kube/admit` package which implements the ExternalAdmissionWebhook for server-side validation of Pilot configuration. This includes a script to enable the workaround described [here](https://github.com/kubernetes/kubernetes/issues/49987#issuecomment-319739227).

This feature requires 1.7.x with [external-admission-webhooks](https://kubernetes.io/docs/admin/extensible-admission-controllers/#external-admission-webhooks). General instructions to enable are [here](https://kubernetes.io/docs/admin/extensible-admission-controllers/#enable-external-admission-webhooks). Specific testing instructions for GKE and Minikube are documented [here](https://github.com/istio/pilot/blob/master/doc/testing.md#create-development-cluster).

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #

Fixes https://github.com/istio/pilot/issues/1066.

**Special notes for your reviewer**:

UPDATED: this PR includes integration with the istio-pilot discovery service. The admission registration is disabled by default since it requires an additional workaround for GKE (see platform/kube/admit/webhook-workaround.sh). 

Users can experiment with this feature on GKE using the e2e pre-submit test with the following sequence of commands.

1) run `bin/e2e.sh -use-admission-webhook`
2) cancel the test after the deployment and test pods are created
3) run `platform/kube/admit/webhook-workaround.sh` and specify the server-name, secret-name, namespace, and port number of the pilot discovery service, e.g.

```bash
platform/kube/admit/webhook-workaround.sh \ 
    --service-name istio-pilot \
    --secret-name pilot-webhook \
    --namespace istio-test-w1g0f \
    --port 443
```

This script converts the istio-pilot discovery service to type=LoadBalancer, waits for ingress IP, creates the workaround external service/endpoint, generates required CA and server certs for webhook, and creates the corresponding k8s secret which is consumed by the admission webhook server.

4) wait until the LoadBalancer IP forwarding rule is propagated in GCP. You can try curl'ing `<ExternalIP>:8000/v1/registration` until a valid response is returned.

5) Create a (in)valid pilot configuration via kubectl (or istioctl). Invalid configuration should be rejected by apiserver with appropriate error message. 

**Release note**:

```release-note
Add the `istio.io/pilot/platform/kube/admit` package which implements the ExternalAdmissionWebhook for server-side validation of Pilot configuration. This includes a script to enable the workaround described [here](https://github.com/kubernetes/kubernetes/issues/49987#issuecomment-319739227).

This feature requires 1.7.x with [external-admission-webhooks](https://kubernetes.io/docs/admin/extensible-admission-controllers/#external-admission-webhooks). General instructions to enable are [here](https://kubernetes.io/docs/admin/extensible-admission-controllers/#enable-external-admission-webhooks). Specific testing instructions for GKE and Minikube are documented [here](https://github.com/istio/pilot/blob/master/doc/testing.md#create-development-cluster).
```
